### ISSUE TITLE
Network Plugin -W option (Prelude.!!: index too large)

### ISSUE THREAD

**Issue by nils-trubkin** [jaor/xmobar#630]
Hi,
### From [docs](https://github.com/jaor/xmobar/blob/master/doc/plugins.org): 
> -W number Bar width
...
Special value: 0. When this parameter is 0, the percentage to display is interpreted as a position in the bar foreground string (given by -f), and the character at that position is displayed.

### My xmobar config (Network Part)
`  Run Network "enp0s3" [ "-t", "<txbar> | <rxbar>"
            , "-L", "125000","-H","87500000"
            , "-n", "#c4e88d"
            , "-h", "#f07180"
            , "-f", "::::::::::"
            , "-b", ".........."
            , "-W", "0" ] 10`
            
### Error (as displayed in the xmobar itself)
`error: Prelude.!!: index too large`

### My guess
This gets triggered at the unsafe !! in [Output.hs, line 227](https://github.com/jaor/xmobar/blob/0b1ad269d109ec46e32910a09339832181e4eb43/src/Xmobar/Plugins/Monitors/Common/Output.hs#L227)
`bfs = if c then [bf !! max 0 (len - 1)] else take len $ cycle bf`
Can't see anything weird there, so probaly just me doing something wrong. Any ideas?

**Comment by jaor** [comment]
i think that what is happening is that the network monitor bars compute
a logarithmic percentage by looking at the (H-L)/W quotient (in
principle, there is no "maximum bandwidth", so it's imposible to compute
a real percentage: we take the low and high thresholds as a hint, and
normalize using the width).  for them, the special value "0" is not
supported.

i've pushed a patch that addresses the problem by choosing an arbitrary
width of 10 in that case: please let me know if that fixes the problem
for you.

(as an aside, in your configuration, having -f equal to a string with the
same character (:) in all positions means that, even if it worked, the
bar would always display a single ':', so it's probably not what you
want anyway.)

**Comment by nils-trubkin** [comment]
Wow that was quick.

I've noticed that I was on the (older) Arch repo version of the xmobar (0.39-137), so now that I pulled down the xmobar-git from AUR (with your patch) it does indeed work. But was it the patch that fixed it or was it just too outdated?

It works exactly as I want it to now. As mentioned in the first post, I've given it some H and L values manually, like 1 Mbit and 0.8 Gbit basically so that should not be an issue.

I do understand that aside you mentioned, I just put it there before posting here. My plan was to use Nerd-Font icons of moon (gradually filled circle) to show the tx and rx of the network. It works perfectly how. One can put any digits 0-9 there too. It's like bar but more minimalistic.

Does that arbitrary '10' for width do anything? Is like if you don't give it the L and H?

**Comment by nils-trubkin** [comment]
![image](https://user-images.githubusercontent.com/32000568/168523367-c5f2f313-443e-4e60-8567-7df7aa968c59.png)

**Comment by jaor** [comment]
On Sun, May 15 2022, Nils Trubkin wrote:

> I've noticed that I was on the (older) Arch repo version of the xmobar
> (0.39-137), so now that I pulled down the xmobar-git from AUR (with
> your patch) it does indeed work. But was it the patch that fixed it or
> was it just too outdated?

no, the problem was there since we introduced the special 0 width.  i
only use it for the battery charge bar, which is not logarithmic, so i
hadn't noticed.

> I do understand that aside you mentioned, I just put it there before
> posting here. My plan was to use Nerd-Font icons of moon (gradually
> filled circle) to show the tx and rx of the network. It works
> perfectly how. One can put any digits 0-9 there too. It's like bar but
> more minimalistic.

aha, that's very nice (you did understand very well, sorry! :). thanks
for sharing! (if you can paste here the relevant config, maybe i could
add it as an example to the docs).

> Does that arbitrary '10' for width do anything? Is like if you don't
> give it the L and H?

it caps the maximum value of the index.  the actual computation is
around here:

  https://github.com/jaor/xmobar/blob/master/src/Xmobar/Plugins/Monitors/Common/Output.hs#L264

**Comment by nils-trubkin** [comment]
> it caps the maximum value of the index.  the actual computation is
> around here:

But does that mean that if I give it a '1234567890abcdef' as the -f string, it should only go up to the '0'? Because, it actually goes up to the letters 'a-f'. I just can not see where this '10' width comes in, does not seem to cap the index in my case.

> if you can paste here the relevant config, maybe I could
> add it as an example to the docs

Yes, absolutely. Currently I'm just a learning and experimenting, so my configs are nothing special, but for this particular case I'm using 1 Mbit as the 'L' and 0.8 Gbit as 'H' since my maximum theoretical bandwidth is 1 Gbit. The numbers are expressed in Bytes, so it's not the most straightforward conversion (division by 8). Maybe it would be nice with an extra option where you specify the unit (b, B, Kb, KB, Mb, MB...) but I don't know if it's hard to implement. My Haskell skills are limited at the moment.

`Run Network "enp0s3" ["-t", "<txbar> 李 <rxbar>"
                                           , "-L", "125000","-H", "100000000"
                                           , "-n", "#c4e88d"
                                           , "-h", "#f07180"
                                           , "-W", "0"
                                           , "-f", "" ] 10`

And since the emojis are all broken here on github, here's a simpler version that should make more visual sense

`Run Network "enp0s3" ["-t", "<txbar> | <rxbar>"
                                           , "-L", "125000","-H", "100000000"
                                           , "-n", "#c4e88d"
                                           , "-h", "#f07180"
                                           , "-W", "0"
                                           , "-f", "0123456789" ] 10`
Config with icons:                                           
![image](https://user-images.githubusercontent.com/32000568/168665578-6eb87c03-3abf-4827-8982-a88786793231.png)

**Comment by nils-trubkin** [comment]
I am currently trying to apply the same kind of config to the other plugins like `Cpu` and `Memory` and they all seem to not really support the `-W 0` option.

### Cpu
Config:
`                     Run Cpu ["-t", " <bar>"
                               , "-L", "4","-H","50"
                               , "-n", "#c3e88d"
                               , "-h", "#f07178"
                               , "-W", "0"
                               , "-f", "" ] 10`

Behaviour: Mostly no prinout, sometimes prints the actual alias `<bar>`

### Memory
Config:
`
                      Run Memory ["-t","龍 <usedbar>"
                                  , "-L", "100","-H","500"
                                  , "-n", "#c3e88d"
                                , "-h", "#f07178"
                                  , "-W", "0"
                                 , "-f", "" ] 10`

Behaviour: Prints out the first index, first icon, but never changes it. Changing the L and H does not help.

**Comment by jaor** [comment]
yes, those plugins (cpu, memory) are using a template parsing method
with which i am unfortunately not familiar (they must date from the
first days of xmobar, when i wasn't yet involved), and they don't
support the -W0 option.  i'll need to study the code with a bit more
pause to add it there.

**Comment by jaor** [comment]
ah, forgot to mention: MultiCpu does support the feature, so perhaps
it'd be a good alternative to Cpu...

**Comment by jaor** [comment]
i just did a quick test for Mem, since looking at the code again i
don't see why it wouldn't work, and with this configuration:

``` haskell
memory' = Memory [ "-t" ,"<usedbar> <usedratio>%"
                 , "-p", "2", "-W", "0","-d", "1", "-f", "0123456789"
                 , "--", "--scale", "1024"] 20
```

is working well for me... right now for instance, i see "2 33%".  this
is a pure percentage bar, so L/H should have no influence whatsoever...

**Comment by nils-trubkin** [comment]
Indeed it works. I guess I've messed my 'testing methodology', sorry for that.
Also yes, the L and H just define the color it seems.

I find it interesting that if both <usedbar> and <used> are present then the L and H will mean threshold of percentage in first case and actual megabytes in second case. I guess it's too niche case to bother, but just something I noticed. On the other hand, if 'Memory' had an 'Alian' argument one could simply run different instances with different thresholds. Probably not something I'd personally want to use though.
